### PR TITLE
add MsgDelegate ica callback

### PR DESCRIPTION
**Pull request by asalzmann**
## What is the purpose of the change

Add `DelegateCallback` to handle delegation state changes.

## Brief Changelog

- Add `DelegateCallback` proto message (stores serialized / deserialized callback args)
- Update `SubmitTx...` functions to take in a callback id, args
- Remove delegate callback logic from ibc_handlers.go
- Add `DelegateCallback` function (+ argument serialization / deserialization functions)
- Call `SubmitTxsStrideEpoch` with "delegate" callback id and marshalled args

## Testing and Verifying

integration tests
```
integration_tests.bats
 ✓ [INTEGRATION-BASIC] address names are correct
 ✓ [INTEGRATION-BASIC] ibc transfer updates all balances
 ✓ [INTEGRATION-BASIC] host zone successfully registered
 ✓ [INTEGRATION-BASIC] liquid stake mints stATOM
 ✓ [INTEGRATION-BASIC] tokens were transferred to GAIA after liquid staking
 ✓ [INTEGRATION-BASIC] tokens on GAIA were staked
 ✓ [INTEGRATION-BASIC] redemption works
 ✓ [INTEGRATION-BASIC] claimed tokens are properly distributed
 ✓ [INTEGRATION-BASIC] rewards are being reinvested (delegated balance increasing)
 ✓ [INTEGRATION-BASIC] exchange rate is updating
 ✓ [NOT-IMPLEMENTED] reinvestment and revenue amounts are correct

11 tests, 0 failures
```
- run through staking flow, observe record deletion

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? no
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? no
  - How is the feature or change documented? not documented

### CODE REVIEW

File: x/stakeibc/keeper/icacallbacks.go [hunk 1]
```diff
@@ -41,15 +48,110 @@ func (c ICACallbacks) AddICACallback(id string, fn interface{}) icacallbackstype
...
+	txMsgData, err := k.GetTxMsgData(ctx, ack)
+	if err != nil {
+		// ack failed, handle here
+		return err
+	}
+	// do we need txMsgData?
+	_ = txMsgData
+
+	// deserialize the args
+	delegateCallback := k.UnmarshalDelegateCallbackArgs(ctx, args)
```
Comment (sampocs) [inline x/stakeibc/keeper/icacallbacks.go:89]: Should we have UnmarshalDelegateCallbackArgs throw an error if it can't unmarshal? Otherwise it will return an empty DelegateCallback{} and I think we'll get a null pointer exception below

**Review by sampocs — COMMENTED** [review]
haven't reviewed this yet, but noticed a couple things when I was referencing it for the undelegated callback

File: x/stakeibc/keeper/icacallbacks.go [hunk 2]
```diff
@@ -41,15 +48,110 @@ func (c ICACallbacks) AddICACallback(id string, fn interface{}) icacallbackstype
...
 // ICACallback Handlers
 // -----------------------------------
 
-func SampleCallback(k Keeper, ctx sdk.Context, packet channeltypes.Packet, ack []byte, args []byte) error {
-	k.Logger(ctx).Info("samplecallback SUCCESS", "packet", packet, "ack", ack, "args", args)
+// ----------------------------------- Delegate Callback ----------------------------------- //
+func (k Keeper) MarshalDelegateCallbackArgs(ctx sdk.Context, delegateCallback types.DelegateCallback) []byte {
+	out, err := proto.Marshal(&delegateCallback)
+	if err != nil {
+		k.Logger(ctx).Error(fmt.Sprintf("UnmarshalDelegateCallbackArgs %v", err.Error()))
```
[the remark this answers is not in the document]
Comment (asalzmann) [reply x/stakeibc/keeper/icacallbacks.go:63]: done

File: x/stakeibc/keeper/icacallbacks.go [hunk 3]
```diff
@@ -41,15 +48,110 @@ func (c ICACallbacks) AddICACallback(id string, fn interface{}) icacallbackstype
...
+	unmarshalledDelegateCallback := types.DelegateCallback{}
+	if err := proto.Unmarshal(delegateCallback, &unmarshalledDelegateCallback); err != nil {
+        k.Logger(ctx).Error(fmt.Sprintf("UnmarshalDelegateCallbackArgs %v", err.Error()))
+	}
+	return unmarshalledDelegateCallback
+}
+
+// QUESTION: Would it be cleaner to pass in ack as a bool (success / failure) here?
+func DelegateCallback(k Keeper, ctx sdk.Context, packet channeltypes.Packet, ack []byte, args []byte) error {
+	k.Logger(ctx).Info("DelegateCallback executing", "packet", packet, "ack", ack, "args", args)
```
[the remark this answers is not in the document]
Comment (asalzmann) [reply x/stakeibc/keeper/icacallbacks.go:78]: removed

File: x/stakeibc/keeper/icacallbacks.go (hunk 1 above)
Comment (asalzmann) [reply x/stakeibc/keeper/icacallbacks.go:89]: good call, done

**Comment by asalzmann** [comment]
@sampocs this is ready for another review, I restructured this so that callbacks take in an Acknowledgement_Result

File: x/stakeibc/keeper/icacallbacks.go [hunk 4]
```diff
@@ -41,15 +46,74 @@ func (c ICACallbacks) AddICACallback(id string, fn interface{}) icacallbackstype
...
+
+	// deserialize the args
+	delegateCallback, err := k.UnmarshalDelegateCallbackArgs(ctx, args)
+	if err != nil {
+		return err
+	}
+	k.Logger(ctx).Info(fmt.Sprintf("DelegateCallback %v", delegateCallback))
+	hostZone := delegateCallback.GetHostZoneId()
+	zone, found := k.GetHostZone(ctx, hostZone)
+	_ = zone
```
Comment (sampocs) [inline x/stakeibc/keeper/icacallbacks.go:92]: nit: remove

File: x/stakeibc/keeper/icacallbacks.go [hunk 5]
```diff
@@ -41,15 +46,74 @@ func (c ICACallbacks) AddICACallback(id string, fn interface{}) icacallbackstype
...
+	if !found {
+		return sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, "host zone not found")
```
Comment (sampocs) [inline x/stakeibc/keeper/icacallbacks.go:94]: nit: log host zone name here

File: x/stakeibc/module_ibc.go [hunk 6]
```diff
@@ -211,7 +186,40 @@ func (im IBCModule) NegotiateAppVersion(
 	return proposedVersion, nil
 }
 
+func (im IBCModule) UnmarshalAck(ctx sdk.Context, acknowledgement []byte) (*channeltypes.Acknowledgement_Result, error) {
+	ack := channeltypes.Acknowledgement_Result{}
+	eventType := "callback"
```
Comment (sampocs) [inline x/stakeibc/module_ibc.go:191]: is this function unique to icacallbacks? I assume it could work with regular ibc acks - could it make sense to pass an eventType as an arg?

File: x/stakeibc/module_ibc.go [hunk 7]
```diff
@@ -211,7 +186,40 @@ func (im IBCModule) NegotiateAppVersion(
...
+	ctx.EventManager().EmitEvent(
+		sdk.NewEvent(
+			eventType,
+			sdk.NewAttribute(sdk.AttributeKeyModule, types.ModuleName),
+			sdk.NewAttribute(types.AttributeKeyAck, fmt.Sprintf("%v", ack)),
```
Comment (sampocs) [inline x/stakeibc/module_ibc.go:196]: is this supposed to be emitted as bytes? I assume so but just double checking

File: x/stakeibc/module_ibc.go [hunk 8]
```diff
@@ -211,7 +186,40 @@ func (im IBCModule) NegotiateAppVersion(
...
+	)
+	err := json.Unmarshal(acknowledgement, &ack)
+	if err != nil {
+		ackErr := channeltypes.Acknowledgement_Error{}
+		err := json.Unmarshal(acknowledgement, &ackErr)
+		if err != nil {
+			ctx.EventManager().EmitEvent(
+				sdk.NewEvent(
+					eventType,
+					sdk.NewAttribute(types.AttributeKeyAckError, ackErr.Error),
```
Comment (sampocs) [inline x/stakeibc/module_ibc.go:207]: need `()` after `Error`?

File: x/stakeibc/module_ibc.go [hunk 9]
```diff
@@ -211,7 +186,40 @@ func (im IBCModule) NegotiateAppVersion(
...
+				),
+			)
+			im.keeper.Logger(ctx).Error("Unable to unmarshal acknowledgement error", "error", err, "data", acknowledgement)
+			return nil, err
+		}
+		im.keeper.Logger(ctx).Error("Unable to unmarshal acknowledgement result", "error", err, "remote_err", ackErr, "data", acknowledgement)
+		return nil, err
```
Comment (sampocs) [inline x/stakeibc/module_ibc.go:214]: nit: logging bytes. Also maybe rephrase the second log. If I'm understanding correctly, this isn't really an unmarshalling error, it's just that ack itself was an error response

File: x/stakeibc/module_ibc.go [hunk 10]
```diff
@@ -211,7 +186,40 @@ func (im IBCModule) NegotiateAppVersion(
...
+	ack := channeltypes.Acknowledgement_Result{}
+	eventType := "callback"
+	ctx.EventManager().EmitEvent(
+		sdk.NewEvent(
+			eventType,
+			sdk.NewAttribute(sdk.AttributeKeyModule, types.ModuleName),
+			sdk.NewAttribute(types.AttributeKeyAck, fmt.Sprintf("%v", ack)),
+		),
+	)
+	err := json.Unmarshal(acknowledgement, &ack)
```
Comment (sampocs) [inline x/stakeibc/module_ibc.go:199]: nit: maybe add a clarifying comment here that the ack can come back as either a success or error type, which is why we first try to unmarshal as a success and then try the error type if that fails

**Review by sampocs — APPROVED** [review]
Looks good! Few nit comments

File: x/stakeibc/keeper/icacallbacks.go (hunk 4 above)
Comment (asalzmann) [reply x/stakeibc/keeper/icacallbacks.go:92]: done

File: x/stakeibc/keeper/icacallbacks.go (hunk 5 above)
Comment (asalzmann) [reply x/stakeibc/keeper/icacallbacks.go:94]: done

File: x/stakeibc/module_ibc.go (hunk 6 above)
Comment (asalzmann) [reply x/stakeibc/module_ibc.go:191]: pulled this event out of `UnmarshalAck`

File: x/stakeibc/module_ibc.go (hunk 7 above)
Comment (asalzmann) [reply x/stakeibc/module_ibc.go:196]: might not make much sense to log this to be honest, removed this line

File: x/stakeibc/module_ibc.go (hunk 8 above)
Comment (asalzmann) [reply x/stakeibc/module_ibc.go:207]: this is just a string attribute, doesn't seem like there's a getter fn

**Comment by sampocs** [comment]
last commit looks good, go ahead and 🚢  when you're ready!
